knowledge-base/adr/0024-url-encoding.md

```diff
@@ -0,0 +1,193 @@
+
+On the other side, the "values of the entity key" of an entity is provided by the user, which might contain special characters.
+One of the users uses `/` as part of a key.
+Therefore, we have to encode the "values of the entity key".
+
+##### Navigation Property Path :thumbsup:
+
+Similar to entity path, only the "values of the navigation property key" are planned to be encoded.
+
+#### OpenAPI
+
+The path pattern should look like the example below:
+
+```
+"/entities/{entityId}"
+```
+
+- path: `/entities` :thumbsdown:
+- path parameter: `{entityId}` :thumbsup:
+
+Similar to OData, only user input is considered when encoding, which is the "path parameters"
+
+### Query
+
+Similar to path section, we also discuss OData and OpenAPI separately.
+
+#### OData
+
+OData allows you to use the following query options:
+
+- `$format=json` :thumbsdown:
+- `$top=1` :thumbsdown:
+- `$skip=1` :thumbsdown:
+- `$orderby=StringProperty` :thumbsdown:
+- `$select=StringProperty` :thumbsdown:
+- `$expand=to_MultiLink` :thumbsup:
+- `$filter=StringProperty eq 'string'` :thumbsup:
+- `$batch` :thumbsdown:
+- `$value` :thumbsdown:
+
+`$filter` is the only one query option that contains user input.
+One spacial case is the `$expand`.
+Normal `$expand` has the same structure as `$select`, while for deep `$expand`, it depends on the inner query option.
+
+#### OpenAPI
+
+The value of the query parameters should be encoded. :thumbsup:
+
+## Alternatives
+
+### Option A: Use `encodeURI` for the whole URL
+
+#### Pros and Cons
+
+##### Pros
+
+- We just need this method ONCE for the whole URL
+- The special characters like `/` (as path delimiter), `&` (as query delimiter) etc. are not encoded. (See [Appendix](#appendix))
+
+##### Cons
+
+- For custom values like custom query, we would like to respect the user input, but the encoding cannot be avoided.
+- When `/` is used as part of the parameter value like `/Entity('a/b/c')`, the `encoode` is not intelligent enough to encode it.
+
+#### Workaround based on this option
+
+- encode `a/b/c` first then apply Option A
+- ask the user to provide encoded value
+
+Both of the 2 proposals will lead to double encoding.
+
+#### Decision
+
+Not accepted, as it does not work for edge cases.
+
+### Option B: Use `encodeURIComponent` for the selected parts
+
+#### Pros and Cons
+
+##### Pros
+
+- We can choose specific components for encoding, which is flexible.
+- It solves the problem of Option A, that `/` can be encoded. (See [Appendix](#appendix))
+- For custom values, the encoding can be skipped.
+
+##### Cons
+
+- We have to call this method multiple times for all the parts that need encoding.
+- We have to be very careful to call `encodeURIComponent('a/b/c')` instead of `encodeURIComponent('/Entity('a/b/c')')`, as the latter one is wrong.
+
+#### Decision
+
+Chosen, as this is a valid solution.
+
+## Decision
+
+Option B is the winner, as Option A does not fit our requirement.
```

Comment:
I think that option A is not an option as it does not solve any of our issues and just doesn't work. I would therefore prefer to follow the standard ADR approach and just describe what we are doing (decision) and give a background + consequences.
If we just have one option, there is less to read and the whole section on option A could just be mentioned in one sentence in the background.

Reply:
I had the same remark. Keep it much shorter:
- Say what we do: encode each user input part
- Say why we do it: encode the full case `host/something/?filter=Foo eq 'abc\def'` can not be solved because the encode method will not encode inside argument where it should but keep the path untouched.